### PR TITLE
[feature] Ascend NPU graph support

### LINKED ISSUE #8030: [Feature] support ACLGraph
### Checklist

- [x] 1. If the issue you raised is not a feature but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
- [x] 2. Please use English, otherwise it will be closed.

### Motivation

# 1. Motivation

After some features had been merged, we could run SGLang on Ascend servers with eager mode. But if we want to get better performance, we need implement ACLGraph or NPUGraph now.

Goals
Goal 1: Define a `NPUGraphRunner` class for SGLang, which provides basic functions and supports llama or Qwen models.

Goal 2: Adapt to TP/DP , GraphTree and dynamic shape scenarios, including memory reuse.

Goal 3: Improve performance based on `torch.compile`, 

# 2. Technical Design

- Workflow Phases

Image

- Key messages

  we have [torch_npu.npu.NPUGraph](https://gitee.com/ascend/pytorch/blob/master/torch_npu/npu/graphs.py#L246), which has similar interfaces and functions to [torch.cuda.CUDAGraph](https://docs.pytorch.org/docs/stable/generated/torch.cuda.CUDAGraph.html)

  Concerning the level of RTS, we can refer to this [document](https://www.hiascend.com/document/detail/zh/canncommercial/81RC1/developmentguide/appdevg/aclcppdevg/aclcppdevg_000519.html). 

# 3. Roadmap

- Phase 1: Basic support for dense model
   
   Implement `NPUGraphRunner` refer to `CUDAGraphRunner`, but we should handle some special case:

   Because we use this [torch_npu.npu_fused_infer_attention_score](https://www.hiascend.com/doc_center/source/zh/Pytorch/60RC2/apiref/apilist/ptaoplist_000787.html) API, which has a host_list input, we have to update its value each time using `torch_npu.npu.NPUGraph.update`. For more details, please refer to [task update](https://www.hiascend.com/document/detail/zh/canncommercial/81RC1/developmentguide/appdevg/aclcppdevg/aclcppdevg_000519.html#ZH-CN_TOPIC_0000002284281029__section4544173013324). 

- Phase 2: Basic support for moe model

### Related resources

_No response_

### PR DESCRIPTION
**Pull request by VDV1985**
## Motivation

NPUGraph can give significant improvement of perfrormance by reducing overhead of operators calling
Closes https://github.com/sgl-project/sglang/issues/8030
Second attempt to enable NPU graph functionality, first one https://github.com/sgl-project/sglang/pull/8027 was reverted

## Modifications

Added NPU graph runner, that inherited from CudaGraphRunner. It is customers request to preserve CudaGraphRunner and not use base class.
cuda_graph_runner.py history preserved. 
Added attention operation that supports by NPUGraph.update()
NPU graph runner uses the same SGLang server options as CUDA graph with combination of 'device="npu"', i.e. "disable_cuda_graph" and enabled by default.
Tests for graph added

_NOTICE_: it's supposed to use "export STREAMS_PER_DEVICE=32" to enable npugraph now. and you won't need it after we release the new version of torch_npu

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

File: python/sglang/srt/model_executor/npu_graph_runner.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+                next_token_logits=output.next_token_logits[: self.raw_num_token],
+                hidden_states=(
+                    output.hidden_states[: self.raw_num_token]
+                    if output.hidden_states is not None
+                    else None
+                ),
+            )
+        else:
+            assert isinstance(output, PPProxyTensors)
+            return PPProxyTensors({k: v[: self.bs] for k, v in output.tensors.items()})
```
[the remark this answers is not in the document]
Comment (VDV1985) [reply python/sglang/srt/model_executor/npu_graph_runner.py:94]: NPUGraph requires call of update in separate thread for performance reasons. Also some performance changes may be introduced in future, so we prefer to keep it as is.

File: python/sglang/srt/layers/attention/ascend_backend.py [hunk 2]
```diff
@@ -167,28 +231,74 @@ def forward_decode(
                 layer, forward_batch.out_cache_loc, k, v
             )
         if not self.use_mla:
-            k_cache = forward_batch.token_to_kv_pool.get_key_buffer(layer.layer_id)
-            v_cache = forward_batch.token_to_kv_pool.get_value_buffer(layer.layer_id)
+            if self.graph_mode:
```
[the remark this answers is not in the document]
Comment (VDV1985) [reply python/sglang/srt/layers/attention/ascend_backend.py:234]: different NPU attentions have different shape requirements for mentioned tensors, so we prefer to keep this as is

File: python/sglang/srt/layers/attention/ascend_backend.py [hunk 3]
```diff
@@ -220,3 +330,6 @@ def forward_decode(
                 out=attn_output,
             )
             return attn_output.view(num_tokens, layer.tp_q_head_num * self.kv_lora_rank)
+
+    def get_cuda_graph_seq_len_fill_value(self):
+        return 0
```
[the remark this answers is not in the document]
Comment (VDV1985) [reply python/sglang/srt/layers/attention/ascend_backend.py]: fixed

File: python/sglang/srt/model_executor/npu_graph_runner.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Run the model with npu graph and torch.compile."""
+
+from __future__ import annotations
+
+import logging
+import threading
+from typing import TYPE_CHECKING
```
[the remark this answers is not in the document]
Comment (VDV1985) [reply python/sglang/srt/model_executor/npu_graph_runner.py]: fixed

**Review by Alcanderian — APPROVED** [review]
This commit satisfies the requirements of minimizing code changes and not affecting history and downstream forks, and can now be approved

**Comment by wangkeya** [comment]
when i test Qwen3-32B input 8k， output500, concurreny is 50 , SGLANG will be hang up 
plus
image

**Comment by VDV1985** [comment]
> when i test Qwen3-32B input 8k， output500, concurreny is 50 , SGLANG will be hang up thus image

@wangkeya could you please share "npu-smi info" output and exact test parameters, i.e. sglang.launch_server params and sglang.bench_serving params (or Engine params if you use other type of testing)?

**Comment by wangkeya** [comment]
> > when i test Qwen3-32B input 8k， output500, concurreny is 50 , SGLANG will be hang up thus image
> 
> @wangkeya could you please share "npu-smi info" output and exact test parameters, i.e. sglang.launch_server params and sglang.bench_serving params (or Engine params if you use other type of testing)?
sglang launch_server
image

bench
image


npu-smi info
image

all threads wait for result
image

**Comment by VDV1985** [comment]
@wangkeya could you please try to set torch_npu environment variable "export STREAMS_PER_DEVICE=32" before running server? Also if you experience issues with memory allocation you can try to increase mem-fraction-static to 0.8 or 0.9.

**Comment by itboyljm** [comment]
Does this MR's merged code support running the DeepSeek model in ACLGraph mode?

**Comment by ping1jing2** [comment]
> Does this MR's merged code support running the DeepSeek model in ACLGraph mode?

I'm sorry to tell you that it's not supported yet. However, we will submit a new PR by 30/08. Thanks for your attention.

**Comment by wangkeya** [comment]
> @wangkeya could you please try to set torch_npu environment variable "export STREAMS_PER_DEVICE=32" before running server? Also if you experience issues with memory allocation you can try to increase mem-fraction-static to 0.8 or 0.9.

thank you very much ,it works

**Comment by wangkeya** [comment]
21d83f8385e58ace4fd18189d326bfe8
@VDV1985 hello, when i test qwen3-235b-a22b, occure this error

**Comment by VDV1985** [comment]
@wangkeya could you please send the error log and server/bench command lines? I suppose you have something like tensor.to("cpu") that causes stream synchronize command that is not supported by graph capturing. Also we have limited support of MOE models with graph at the moment and working on full support.

**Comment by VDV1985** [comment]
@wangkeya , also you can look at this PR https://github.com/sgl-project/sglang/pull/10013. May be it can help.

**Comment by wangkeya** [comment]
> #10013

image
image
i will try this pr 10013

**Comment by wangkeya** [comment]
> > #10013
> 
> image image i will try this pr 10013

@VDV1985 2*910B, when request concurrency is 7, sglang will be hang up。export STREAMS_PER_DEVICE=32 already set 

image

**Comment by ping1jing2** [comment]
> > > #10013
> > 
> > 
> > image image i will try this pr 10013
> 
> @VDV1985 2*910B, when request concurrency is 7, sglang will be hang up。export STREAMS_PER_DEVICE=32 already set
> 
> image

@wangkeya please open a new issue and assign me, let us help you to solve the problem.

**Comment by wangkeya** [comment]
> > > > #10013
> > > 
> > > 
> > > image image i will try this pr 10013
> > 
> > 
> > @VDV1985 2*910B, when request concurrency is 7, sglang will be hang up。export STREAMS_PER_DEVICE=32 already set
> > image
> 
> @wangkeya please open a new issue and assign me, let us help you to solve the problem.

thank you , new issue is https://github.com/sgl-project/sglang/issues/10091
